lib/crypto/src/bigint.rs

```diff
@@ -0,0 +1,277 @@
+    + BitOr<Self, Output = Self>
+    + for<'a> BitOr<&'a Self, Output = Self>
+    + Shr<usize, Output = Self>
+    + ShrAssign<usize>
+    + Shl<usize, Output = Self>
+    + ShlAssign<usize>
+{
+    /// Number of `usize` limbs representing `Self`.
+    const NUM_LIMBS: usize;
+
+    /// Returns true if this number is odd.
+    /// # Example
+    ///
+    /// ```
+    /// use openzeppelin_crypto::bigint::{BigInteger, crypto_bigint::U64};
+    ///
+    /// let mut one = U64::from(1u64);
+    /// assert!(one.is_odd());
+    /// ```
+    fn is_odd(&self) -> bool;
+
+    /// Returns true if this number is even.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use openzeppelin_crypto::bigint::{BigInteger, crypto_bigint::U64};
+    ///
+    /// let mut two = U64::from(2u64);
+    /// assert!(two.is_even());
+    /// ```
+    fn is_even(&self) -> bool;
+
+    /// Returns true if this number is zero.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use openzeppelin_crypto::bigint::{BigInteger, crypto_bigint::U64};
+    ///
+    /// let mut zero = U64::from(0u64);
+    /// assert!(zero.is_zero());
+    /// ```
+    fn is_zero(&self) -> bool;
+
+    /// Compute the minimum number of bits needed to encode this number.
+    /// # Example
+    /// ```
+    /// use openzeppelin_crypto::bigint::{BigInteger, crypto_bigint::U64};
+    ///
+    /// let zero = U64::from(0u64);
+    /// assert_eq!(zero.num_bits(), 0);
+    /// let one = U64::from(1u64);
+    /// assert_eq!(one.num_bits(), 1);
+    /// let max = U64::from(u64::MAX);
+    /// assert_eq!(max.num_bits(), 64);
+    /// let u32_max = U64::from(u32::MAX as u64);
+    /// assert_eq!(u32_max.num_bits(), 32);
+    /// ```
+    fn num_bits(&self) -> usize;
+
+    /// Compute the `i`-th bit of `self`.
+    /// # Example
+    ///
+    /// ```
+    /// use openzeppelin_crypto::bigint::{BigInteger, crypto_bigint::U64};
+    ///
+    /// let mut one = U64::from(1u64);
+    /// assert!(one.get_bit(0));
+    /// assert!(!one.get_bit(1));
+    /// ```
+    fn get_bit(&self, i: usize) -> bool;
+}
+
+impl<const N: usize> BigInteger for Uint<N> {
+    const NUM_LIMBS: usize = N;
+
+    fn is_odd(&self) -> bool {
+        <Uint<N> as Integer>::is_odd(self).into()
+    }
+
+    fn is_even(&self) -> bool {
+        <Uint<N> as Integer>::is_even(self).into()
+    }
+
+    fn is_zero(&self) -> bool {
+        <Uint<N> as Zero>::is_zero(self).into()
+    }
+
+    fn num_bits(&self) -> usize {
+        self.bits()
+    }
+
+    fn get_bit(&self, i: usize) -> bool {
+        self.bit(i).into()
+    }
+}
+
+impl<const N: usize> BitIteratorBE for Uint<N> {
+    fn bit_be_iter(&self) -> impl Iterator<Item = bool> {
+        self.as_words().iter().rev().flat_map(Word::bit_be_iter)
+    }
+}
+
+/// Parse a number from a string in a given radix.
+///
+/// I.e., convert string encoded integer `s` to base-`radix` number.
+#[must_use]
+pub const fn from_str_radix<const LIMBS: usize>(
+    s: &str,
```

Comment:
Why not `AsRef<str>`?

Reply:
This function also have a `const` context, and we can not call a constant `AsRef::as_ref` trait method in `const` function context.
And the hole point of this imperative implementation is: the function can be evaluated at compile time.